socorro/signature/__main__.py

```diff
@@ -0,0 +1,260 @@
+
+    def warning(self, line):
+        """Prints out a warning line to stderr
+
+        :arg str line: the line to print to stderr
+
+        """
+        print('WARNING: %s' % line, file=sys.stderr)
+
+    def data(self, crash_id, old_sig, new_sig, notes):
+        """Outputs a data point
+
+        :arg str crash_id: the crash id for the signature generated
+
+        :arg str old_sig: the old signature retrieved in the processed crash
+
+        :arg str new_sig: the new generated signature
+
+        :arg list notes: any processor notes
+
+        """
+        pass
+
+
+class TextOutput(OutputBase):
+    def data(self, crash_id, old_sig, new_sig, notes):
+        print('Original: %s' % old_sig)
+        print('New:      %s' % new_sig)
+        print('Same?:    %s' % (old_sig == new_sig))
+
+        if notes:
+            print('Notes:    (%d)' % len(notes))
+            for note in notes:
+                print('          %s' % note)
+
+
+class CSVOutput(OutputBase):
+    def __enter__(self):
+        self.out = csv.writer(sys.stdout, quoting=csv.QUOTE_ALL)
+        self.out.writerow(['crashid', 'old', 'new', 'same?', 'notes'])
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.out = None
+
+    def data(self, crash_id, old_sig, new_sig, notes):
+        self.out.writerow([crash_id, old_sig, new_sig, str(old_sig == new_sig), notes])
+
+
+def fetch(endpoint, crash_id, api_token=None):
+    kwargs = {
+        'params': {
+            'crash_id': crash_id
+        }
+    }
+    if api_token:
+        kwargs['headers'] = {
+            'Auth-Token': api_token
+        }
+
+    return requests.get(API_URL + endpoint, **kwargs)
+
+
+def main(args):
+    """Takes crash data via args and generates a Socorro signature
+
+    """
+    parser = argparse.ArgumentParser(description=DESCRIPTION, epilog=EPILOG)
+    parser.add_argument(
+        '-v', '--verbose', help='increase output verbosity', action='store_true'
+    )
+    parser.add_argument(
+        '--format', help='specify output format: csv, text (default)'
+    )
+    parser.add_argument(
+        'crashids', metavar='crashid', nargs='+', help='crash id to generate signatures for'
+    )
+    args = parser.parse_args()
+
+    if args.format == 'csv':
+        outputter = CSVOutput
+    else:
+        outputter = TextOutput
+
+    if args.verbose:
+        logging_level = 'DEBUG'
+    else:
+        logging_level = 'INFO'
+
+    api_token = os.environ.get('SOCORRO_API_TOKEN', '')
+
+    setup_logging(logging_level)
+
+    generator = SignatureGenerator(debug=args.verbose)
+
+    with outputter() as out:
+        for crash_id in args.crashids:
+            resp = fetch('/RawCrash/', crash_id, api_token)
+            if resp.status_code == 404:
+                out.warning('%s: does not exist.' % crash_id)
+                continue
+            if resp.status_code == 429:
+                out.warning('API rate limit reached. %s' % resp.content)
```

Comment:
The chance of hitting the limit is quite small because of the latency of downloading is high and thus you can't make that many requests. One solution, so to say, is to mention in the logging warning that this is alleviated by using an API token.